content/en/blog/_posts/2025-XX-XX-nftables-kube-proxy/index.md

```diff
@@ -0,0 +1,238 @@
+services and endpoints change.
+
+With both iptables and nftables, the total size of the "ruleset" as a
+whole (actual "rules", plus associated data) is **O(n)** in the combined
+number of services and endpoints. Originally, the iptables backend would
+rewrite every rule on every update, and with tens of thousands of
+services, this could grow to be hundreds of thousands of iptables rules.
+Starting in Kubernetes 1.26, we began improving kube-proxy so that it
+could skip updating _most_ of the unchanged rules in each update, but
+the limitations of `iptables-restore` as an API meant that it was still
+always necessary to send an update that's **O(n)** in the number of
+services (though with a noticeably smaller constant than it used to be).
+Even with those optimizations, it can still be necessary to make use of
+kube-proxy's `minSyncPeriod` config option to ensure that it doesn't
+spend every waking second trying to push iptables updates.
+
+The nftables APIs allow for doing much more incremental updates, and
+when kube-proxy in nftables mode does an update, the size of the
+update is only **O(n)** in the number of _changed_ services and
+endpoints. The fact that the nftables API allows each nftables-using
+component to have its own private table also means that there is no
+global lock contention between components like with iptables. As a
+result, kube-proxy's nftables updates can be done much more
+efficiently than with iptables.
+
+(Unfortunately I don't have cool graphs for this part.)
+
+## Why _not_ NFTables?
+
+All that said, there are a few reasons why you might not want to jump
+right into using the nftables backend for now.
+
+First, while kube-proxy's nftables mode behaves very similarly to the
+iptables mode at a high level (since both APIs speak to the same
+underlying kernel subsystem), it's still possible that some existing
+network plugins that work with kube-proxy in iptables mode might not
+work right with it in nftables mode; you should consult the
+documentation (or forums, bug tracker, etc.) for your network plugin to
+see if it supports nftables mode. (Note that some plugins may not need
+any changes; as long as they don't try to directly interact with or
+override kube-proxy's iptables rules, they shouldn't care whether
+kube-proxy is using iptables or nftables.)
+
+Second, the nftables mode will not work on older Linux distributions;
+currently it requires a 5.13 or newer kernel. Additionally, because of
+bugs in early versions of the `nft` command line tool, you should not
+run kube-proxy in nftables mode on nodes that have a pre-"1.0.0" version
+of `nft` in the host filesystem (or else kube-proxy's use of nftables
+may interfere with other uses of nftables on the system).
+
+Finally, kube-proxy in nftables mode is intentionally not 100%
+compatible with kube-proxy in iptables mode, because some of iptables
+mode's legacy behaviors are inherently less efficient or less secure.
```

Comment:
I would phrase it differently as this reads negative, and you explain this always very well as "since iptables-mode was GA for a long time, there are certain behaviors we wanted to change but we couldnt because they will break compatibility, however, nftables mode allowed us to make this modifications to improve this behaviors at the cost of breaking compatibility between different proxy modes"